File: labs/part2/zad49.hs
-- Funkcja obliczająca sumę dzielników wszystkich liczb od 1 do maxN
calculateSigmas :: Int -> [Int]
calculateSigmas maxN = map sigma [1..maxN]
  where
    sigma n = sum [i | i <- [1..n], n `mod` i == 0]

-- Funkcja obliczająca wartości pn dla n = 1 do maxN
calculatePn :: Int -> [Int]
calculatePn maxN = pn
  where
    sigmas = calculateSigmas maxN
    pn = 1 : [sum [(sigmas !! (j - 1)) * (pn !! (n - j)) | j <- [1..n]] `div` n | n <- [1..maxN]]

-- Wywołanie funkcji dla n = 1 do 100
main :: IO ()
main = print $ take 100 $ calculatePn 100

File: labs/part2/zad49.java
public class zad49 {
    public static void main(String[] args) {
        int maxN = 100;
        long[] pn = calculatePn(maxN);

        // Wyświetlenie pierwszych 10 wartości dla demonstracji
        for (int i = 1; i <= 100; i++) {
            System.out.println("p(" + i + ") = " + pn[i]);
        }
    }

    private static long[] calculatePn(int maxN) {
        long[] pn = new long[maxN + 1];
        int[] sigmaValues = calculateSigmaUpTo(maxN);

        pn[0] = 1; // p0 = 1 jako przypadek bazowy

        for (int n = 1; n <= maxN; n++) {
            pn[n] = 0;
            for (int j = 1; j <= n; j++) {
                pn[n] += sigmaValues[j] * pn[n - j];
            }
            pn[n] /= n;
        }

        return pn;
    }

    private static int[] calculateSigmaUpTo(int maxN) {
        int[] sigmaValues = new int[maxN + 1];
        for (int n = 1; n <= maxN; n++) {
            for (int i = n; i <= maxN; i += n) {
                sigmaValues[i] += n;
            }
        }
        return sigmaValues;
    }
}
